Userland/Utilities/strings.cpp

```diff
@@ -0,0 +1,88 @@
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/CharacterTypes.h>
+#include <LibCore/ArgsParser.h>
+#include <LibCore/System.h>
+#include <LibMain/Main.h>
+#include <unistd.h>
+
+ErrorOr<int> serenity_main(Main::Arguments arguments)
+{
+    TRY(Core::System::pledge("stdio rpath"));
+
+    Vector<StringView> paths;
+
+    Core::ArgsParser args_parser;
+    args_parser.set_general_help("Write the sequences of printable characters in files or pipes to stdout.");
+    args_parser.add_positional_argument(paths, "File path", "path", Core::ArgsParser::Required::No);
+    args_parser.parse(arguments);
+
+    Vector<int> fds;
+    if (paths.is_empty()) {
+        TRY(fds.try_append(STDIN_FILENO));
+    } else {
+        for (auto const& path : paths) {
+            int fd;
+            if (path == "-") {
+                fd = 0;
+            } else {
+                auto fd_or_error = Core::System::open(path, O_RDONLY);
+                if (fd_or_error.is_error()) {
+                    warnln("Failed to open {}: {}", path, fd_or_error.error());
+                    continue;
+                }
+                fd = fd_or_error.release_value();
+            }
+            TRY(fds.try_append(fd));
+        }
+    }
+
+    TRY(Core::System::pledge("stdio"));
+
+    Array<u8, 32768> buffer;
+    Vector<u8> output_characters;
+    for (auto& fd : fds) {
+        for (;;) {
+            auto buffer_span = buffer.span();
+            auto nread = TRY(Core::System::read(fd, buffer_span));
+            if (nread == 0)
+                break;
+            buffer_span = buffer_span.trim(nread);
+            while (!buffer_span.is_empty()) {
+                int processed_characters = 0;
+                for (u8 ch : buffer_span) {
+                    ++processed_characters;
+                    if (is_ascii_printable(ch) || ch == '\t')
+                        output_characters.append(ch);
+                    else {
+                        if (output_characters.size() < 4)
+                            output_characters.clear();
+                        break;
+                    }
+                }
+                if (output_characters.size() >= 4) {
+                    bool should_output_characters = false;
+                    for (u8 ch : output_characters) {
+                        if (is_ascii_printable(ch) && !is_ascii_space(ch)) {
+                            should_output_characters = true;
+                            break;
+                        }
+                    }
+                    if (should_output_characters) {
+                        TRY(Core::System::write(STDOUT_FILENO, output_characters.span()));
+                        outln("");
+                    }
+                    output_characters.clear();
+                }
+                buffer_span = buffer_span.slice(processed_characters);
+            }
+        }
+        TRY(Core::System::close(fd));
+    }
+
+    return 0;
```

Comment:
This must return different error codes when different kind of issues occur:

- Error code for no input provided
- Error code for invalid arguments
- Error code for IO error(s) if any

See https://github.com/gittup/binutils/blob/gittup/binutils/strings.c

Reply:
We should reference POSIX, not GNU as our source of truth. https://pubs.opengroup.org/onlinepubs/9699919799/utilities/strings.html

Reply:
> * Error code for no input provided

POSIX specifies "If no file operand is specified, the strings utility shall read from the standard input." which matches the current behavior.

> * Error code for invalid arguments

Invalid options display the message "Unrecognized option <option>" and the program returns error code 1.

> * Error code for IO error(s) if any

This is implemented in [7784c1d](https://github.com/SerenityOS/serenity/pull/16013/commits/7784c1dd97530ae67ae2ff14ed2aa277acfbf7eb).